### DISCUSSION TITLE
How to Modify Exchange Rules in Metropolis Sampling for 1/2 Spin Fermion Systems in NetKet?

### DISCUSSION THREAD

**Discussion by Mio826** [netket/netket#1962] [Q&A]
(This is a question due to lack of experience)
When doing Metropolis sampling on a 1/2 spin fermion system, the exchange rule seems to only allow exchange of electrons within a total spin conserved subspace according to this file:  **netket\sampler\rules\exchange.py**
![图片](https://github.com/user-attachments/assets/7f454c4a-8b70-4548-b685-e5b95d669284).
If I faile to understand the code right, please advise me, otherwise, how should I change this rule, to allow exchange of electrons with different spin? (besides inputing another clusters or graph including two subspace)

**Comment by PhilipVinc** [comment]
See https://netket.readthedocs.io/en/latest/api/_generated/samplers/netket.sampler.MetropolisExchange.html#netket.sampler.MetropolisExchange

> If you are working with systems where the number of nodes in the physical lattice does not match the number of degrees of freedom, you must be careful!
>
>A typical example is a system of Spin-1/2 fermions on a lattice with N sites, where the first N degrees of freedom correspond to the spin down degrees of freedom and the next N degrees of freedom correspond to the spin up degrees of freedom.
>
>In this case, you tipically want to exchange only degrees of freedom of the same type. A simple way to achieve this is to double the graph:
>
> ```python
>import netket as nk
>g = nk.graph.Square(5)
>hi = nkx.hilbert.SpinOrbitalFermions(g.n_nodes, s=0.5)
>
>exchange_graph = nk.graph.disjoint_union(g, g)
>print("Exchange graph size:", exchange_graph.n_nodes)
>
>sa = nk.sampler.MetropolisExchange(hi, graph=exchange_graph, d_max=1)
>```
